sort javascript makers, add ternlint maker

davidosomething opened the pull request:
- FIX: #287 -- add tern-lint support as the `ternlint` maker
- CHANGED: sort javascript makers alphabetically (there's no logic currently)
- CHANGED: add `ternlint` maker to default makers

blocked by https://github.com/angelozerr/tern-lint/issues/76

bounceme commented:
Honestly, i consider tern-lint as dangerous. To have any kind of accuracy it needs to look at all the dependencies in your project. This is absolutely impractical in most projects and the process will continue for minutes/ until the node stack is blown. Consider how easy people will forget that the process is running? what about running synchronously? Many people don't seem to know about C-c and would just kill vim losing their work. I would also have some concern about the sheer size of output, massive in many cases ( because the dependency parse ). could viml actually handle it all?

davidosomething commented:
**I'm going to close this PR** because  
- tern-lint is indeed not good enough to be useful
- neomake needs more features to use it well

blocked by #428, #543, https://github.com/angelozerr/tern-lint/pull/72

I have a test case here, where I run .tern-lint on 836k of JS split amongst 124 files
https://gist.github.com/davidosomething/1b6e06bc2d263547cd9a4ce4ac13214e  

- There are a ton of false positives, but that's something to update in tern-lint
- The total time is 1.2seconds to run tern-lint (`node_modules` and vendor stuff excluded) so speed is not a factor when configured correctly
- stack size and files opened IS a factor for large projects

I'll probably update this in the far future to not enable tern by default (and even though it is on, user would need to have both npm installed tern-lint globally and configured a .tern-project to run it). Then users would have to manually use `:Neomake tern` to run it.

Some notes addressing your points above and if anyone wants to take over this:

- tern by default does not run with the modules plugin so will not look at your dependencies
- with the various modules plugins on, you can also specify `dontLoad` to not load modules (in which case only files included in loadEagerly in `.tern-project` will be checked)
    - A good addition for tern-lint is as you've done in https://github.com/angelozerr/tern-lint/pull/72/files to populate loadEagerly based on the CLI file arg, otherwise it runs against all files loaded eagerly in the tern project or already loaded in the tern server.

- the tern server closes itself after 5 minutes be default
- when running tern-lint synchronously (i.e. the case of neomake in Vim7.4) it will indeed block the vim process -- as will all other make programs, and the person using neomake in vim7.4 should know about that already
- if the node stack is blown (i have yet to encounter this), the process will probably exit with error, in which case neovim just reports an error with the maker
- if the files opened limit is reached (OS limit) everything will shit the bed (problem to address in tern-lint, again I like PR 72 in tern-lint)

- re:https://github.com/angelozerr/tern-lint/pull/72 @angelozerr was under the impression that you'd be passing in a stream as you type (on-the-fly linting as mentioned in other neomake PRs) to the tern-lint cli. This is not the case with neomake -- it passes in a filename when run with `:Neomake`, and tern-lint cli will run it against the file in the filesystem. No point in running neomake on `InsertChanged` (until #543)
    - The server approach mentioned is a good way to go about it -- depends on #428
